README.md
<!--
    Developer: Sikandar Ali Bhojawat
    Github: https://github.com/SikandarAliBhojawat
-->

Create a table in very easy way with simple and clean UI. and also customize it easily.

## Table Example

<img src="./images/image1.jpg" width="100" alt="Image 1">
<img src="./images/image2.jpg" width="100" alt="Image 2">
<img src="./images/image3.jpg" width="100" alt="Image 3">
<img src="./images/image4.jpg" width="100" alt="Image 4">

## 🚀 Features

- **🎨 Beautiful UI**: Create stunning tables with ease, thanks to a sleek and modern design.

- **🛠 Clean Code**: Write maintainable code effortlessly, as StylishTable encourages clean and readable code practices.

- **🚀 Easy Integration**: Quickly integrate stylish tables into your Flutter applications with minimal setup.

- **🌐 Responsive Design**: Ensure a seamless user experience across different screen sizes, making your tables look great on any device.

- **🎨 Customization**: Tailor the look and feel of your tables using a variety of styling options, including colors, fonts, and borders.

## 📦 Installation

Add the following to your `pubspec.yaml` file:

```yaml
dependencies:
  stylish_table: ^1.0.0
```
OR

You can run from terminal by this command

```dart
$ flutter pub add stylish_table 
```

Then, run:

```dart
$ flutter pub get
```

## Usage
Here some example code to insert table in your code. First of all you need to import package in your code.

```dart
import 'package:stylish_table/stylish_table.dart';
```
After import you can use code easily.

Simple implementation of stylish_table.
```dart
  StylishTable(
    tableData: tableData,
    titleData: titleData,
    titleBackgroundColor: Colors.deepPurple,
  ),
```

Full Modified Code Here:
```dart
  StylishTable(
    tableData: tableData,
    titleData: titleData,
    titleBackgroundColor: Colors.deepPurple,
    rowTextColor: Colors.white,
    tableBodyHeight: double.infinity,
    tableBodyWidth: double.infinity,
    tableRowHeight: 45,
    titleRowHeight: 45,
    titleTextColor: Colors.white,
    firstRowColor: Colors.green,
    secondRowColor: Colors.black38,
  ),
```
You need to make data in list or collection in this type example.

Table's title bar variable example here :
```dart
  List<String> titleData = [
    "ID",
    "BetAmount",
    "Win/Loss",
    "Result",
  ];
```
And Table's data variable and collection example here:
```dart
final Map<String, Map<String, String>> tableData = {
    'row1' : { 'id' : 'ID1', 'betAmount' : '500', 'winLoss' : '+1500', 'result' : 'Win' },
    'row2' : { 'id' : 'ID2', 'betAmount' : '100', 'winLoss' : '-100', 'result' : 'Loss' },
    'row3' : { 'id' : 'ID3', 'betAmount' : '300', 'winLoss' : '-100', 'result' : 'Loss' },
    'row4' : { 'id' : 'ID4', 'betAmount' : '50', 'winLoss' : '+500', 'result' : 'Win' },
    'row5' : { 'id' : 'ID5', 'betAmount' : '700', 'winLoss' : '-700', 'result' : 'Loss' },
    'row6' : { 'id' : 'ID6', 'betAmount' : '500', 'winLoss' : '+1500', 'result' : 'Win' },
  };
```
Only use data in this form to avoid errors. otherwise you face some erros or issues.

## 📜 License
This project is licensed under the MIT License - see the LICENSE file for details.

## 🙌 Acknowledgments
Special thanks to the Flutter community for inspiration and support.


CHANGELOG.md
## 0.0.1

* Initial version : 0.0.1


test/stylish_table_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:stylish_table/stylish_table.dart';

void main() {
  test('adds one to input values', () {
    const StylishTable(tableData: {}, titleData: [], titleBackgroundColor: Colors.black,);
  });
}


lib/stylish_table.dart
library stylish_table;

import 'package:flutter/material.dart';

class StylishTable extends StatefulWidget {
  final List<String> titleData; // Make titleData optional
  final Map<String, Map<String, String>> tableData;
  final Color titleTextColor, firstRowColor, secondRowColor, rowTextColor, titleBackgroundColor;
  final double tableBodyHeight, tableBodyWidth, titleRowHeight, tableRowHeight;

  const StylishTable({
    Key? key,
    required this.tableData,
    required this.titleData,
    required this.titleBackgroundColor,
    this.firstRowColor = Colors.black26,
    this.secondRowColor = Colors.black38,
    this.rowTextColor = Colors.white,
    this.titleTextColor = Colors.white,
    this.tableRowHeight = 45,
    this.titleRowHeight = 45,
    this.tableBodyHeight = double.infinity,
    this.tableBodyWidth = double.infinity,
  }) : super(key: key);

  @override
  State<StylishTable> createState() => _StylishTableState();
}

class _StylishTableState extends State<StylishTable> {
  @override
  Widget build(BuildContext context) {
    //List<String> reversedKeys = widget.tableData.keys.toList().reversed.toList();
    return Container(
      height: widget.tableBodyHeight > 0
          ? widget.tableBodyHeight
          : MediaQuery.of(context).size.height,
      width: widget.tableBodyWidth > 0
          ? widget.tableBodyWidth
          : MediaQuery.of(context).size.width,
      child: SafeArea(
        child: Column(
          children: [
            Container(
              height: widget.titleRowHeight,
              width: widget.tableBodyWidth > 0
                  ? widget.tableBodyWidth
                  : MediaQuery.of(context).size.width,
              decoration: BoxDecoration(
                color: widget.titleBackgroundColor,
              ),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: [
                  ListView.builder(
                      shrinkWrap: true,
                      itemCount: widget.titleData.length,
                      itemExtent: MediaQuery.of(context).size.width *
                          (1 / widget.titleData.length),
                      scrollDirection: Axis.horizontal,
                      itemBuilder: (context, index) {
                        return Center(
                            child: Text(
                          widget.titleData[index],
                          style: TextStyle(
                              color: widget.titleTextColor,
                              fontWeight: FontWeight.bold),
                        ));
                      }),
                ],
              ),
            ),
            Expanded(
              child: ListView.builder(
                  itemCount: widget.tableData.length,
                  shrinkWrap: true,
                  itemBuilder: (context, indexOuter) {
                    var row = widget.tableData[
                        widget.tableData.entries.elementAt(indexOuter).key];
                    //print('${row?.length}');
                    return Container(
                      height: widget.tableRowHeight,
                      width: widget.tableBodyWidth > 0
                          ? widget.tableBodyWidth
                          : MediaQuery.of(context).size.width,
                      color: indexOuter % 2 == 0
                          ? (widget.firstRowColor != Colors.black26 ?  widget.firstRowColor : widget.titleBackgroundColor.withOpacity(0.4))
                          : (widget.secondRowColor != Colors.black38 ?  widget.secondRowColor : widget.titleBackgroundColor.withOpacity(0.6)),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceAround,
                        children: [
                          ListView.builder(
                              shrinkWrap: true,
                              itemCount: widget.titleData.length == row!.length
                                  ? widget.titleData.length
                                  : row.length,
                              itemExtent: MediaQuery.of(context).size.width *
                                  (1 / widget.titleData.length),
                              scrollDirection: Axis.horizontal,
                              itemBuilder: (context, indexInner) {
                                var id = row.entries;
                                //print('Hello this is row items :  ${id?.elementAt(indexInner).value}');
                                return Center(
                                    child: Text(
                                  id.elementAt(indexInner).value,
                                  style: TextStyle(
                                      color:
                                          widget.rowTextColor,
                                      fontWeight: FontWeight.bold),
                                ));
                              }),
                        ],
                      ),
                    );
                  }),
            )
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:stylish_table/stylish_table.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key,});


  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<String> titleData = [
    "ID",
    "BetAmount",
    "Win/Loss",
    "Result",
  ];

  final Map<String, Map<String, String>> tableData = {
    'row1' : { 'id' : 'ID1', 'betAmount' : '500', 'winLoss' : '+1500', 'result' : 'Win' },
    'row2' : { 'id' : 'ID2', 'betAmount' : '100', 'winLoss' : '-100', 'result' : 'Loss' },
    'row3' : { 'id' : 'ID3', 'betAmount' : '300', 'winLoss' : '-100', 'result' : 'Loss' },
    'row4' : { 'id' : 'ID4', 'betAmount' : '50', 'winLoss' : '+500', 'result' : 'Win' },
    'row5' : { 'id' : 'ID5', 'betAmount' : '700', 'winLoss' : '-700', 'result' : 'Loss' },
    'row6' : { 'id' : 'ID6', 'betAmount' : '500', 'winLoss' : '+1500', 'result' : 'Win' },
    'row7' : { 'id' : 'ID7', 'betAmount' : '100', 'winLoss' : '-100', 'result' : 'Loss' },
    'row8' : { 'id' : 'ID8', 'betAmount' : '300', 'winLoss' : '-100', 'result' : 'Loss' },
    'row9' : { 'id' : 'ID9', 'betAmount' : '50', 'winLoss' : '+500', 'result' : 'Win' },
    'row10' : { 'id' : 'ID10', 'betAmount' : '700', 'winLoss' : '-700', 'result' : 'Loss' },
    'row11' : { 'id' : 'ID11', 'betAmount' : '500', 'winLoss' : '+1500', 'result' : 'Win' },
    'row12' : { 'id' : 'ID12', 'betAmount' : '100', 'winLoss' : '-100', 'result' : 'Loss' },
    'row13' : { 'id' : 'ID13', 'betAmount' : '300', 'winLoss' : '-100', 'result' : 'Loss' },
    'row14' : { 'id' : 'ID14', 'betAmount' : '50', 'winLoss' : '+500', 'result' : 'Win' },
    'row15' : { 'id' : 'ID15', 'betAmount' : '700', 'winLoss' : '-700', 'result' : 'Loss' },
    'row16' : { 'id' : 'ID16', 'betAmount' : '700', 'winLoss' : '-700', 'result' : 'Loss' },
    'row17' : { 'id' : 'ID17', 'betAmount' : '700', 'winLoss' : '-700', 'result' : 'Loss' },
  };

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: StylishTable(
        tableData: tableData,
        titleData: titleData,
        titleBackgroundColor: Colors.black,
        titleTextColor: Colors.white,
        rowTextColor: Colors.white70,
        // tableBodyHeight: double.infinity,
        // tableBodyWidth: double.infinity,
        // tableRowHeight: 45,
        // titleRowHeight: 45,
        // firstRowColor: Colors.green,
        // secondRowColor: Colors.black38,
      ),
    );
  }
}